A3C barebones for cup picking

beduffy opened the pull request:
Very clean ikrostrikov A3C applied to AI2Thor gym task of cup picking. Can learn pretty fast to do this task well. And achieves optimal in ~100-400k steps i.e. spamming cup picking.

- Good argparsing.
- async vs sync mode.
- Finally using test file properly. Removed plotting since barebones
- Removed reward clamping
- Reduce duplication and many comment variation lines
- Combined old A3C branch into this one for reference and merging best of both worlds
- Open question about channel order in ai2thor and whether to make the change in this PR? @fgtoralesch 
- Open question about atari option for encouraging modularity? @fgtoralesch 
- Closing env not working ideally

beduffy committed:
Created beautiful gym wrapper for AI2Thor with example random agent

All ugliness is self-contained

beduffy committed:
Adapted and got ikostrikov's pytorch-a3c working on AI2Thor wrapper

grayscale 1 channel input, resolution (64, 64)

beduffy committed:
Added test_train_py_sync for synchronous testing of environment e.g.

Going forward 20 steps and watching gradients in debug mode.
Reset log statement

beduffy committed:
Added train_synchronous.py for easily testing and using pdb for debugging

Still works well regardless for training

beduffy committed:
A3C training well with avgavg monitoring and fixed total reward in episode monitoring

Added constant negative reward to encourage speed.

beduffy committed:
Added max episode length param to wrapper, self.done, better resetting, big reward

beduffy committed:
Added 4 files from Ilya Kostrikov's pytorch-a3c (https://github.com/ikostrikov/pytorch-a3c)

main, my_optim, train and main with Adapted from at top. No other changes

beduffy committed:
Added atari envs (could have to option to try both in future) and test.py (didn't realise it slept)

beduffy committed:
opening ai2thor environment instead of atari (but left commented).
Still crashing on shape error for conv
default 1 processes. Tried close env.

beduffy committed:
Merge branch 'A3C' into A3C_barebones
to get main ideas+changes from when I got A3C working in September and integrate into new wrapper

beduffy committed:
Added synchronous option to main this time to avoid duplication.

Works in this mode. passing max_episode_length to config_dict
some todos. Added 128x128 resolution model and unsqueeze if 3 dimension check

beduffy committed:
Got test working by making state float. Added sync and async short versions.

Added test-sleep-time arg, very important for processing power
add and comment out test() in sync mode

beduffy committed:
Added many changes to train.py. For example. No plotting,

cleaner episode finished log. Removed reward clamping. Monitoring with 4 lists
Total length

beduffy committed:
Added async back and added config_dict to args

which is to be passed to all processes and read by them to create env
added todo

beduffy committed:
Deleted 7 old A3C files from another branch

One or two things could be done in the future from them e.g. plotting
and

beduffy committed:
Added print and comment. Created PR.

beduffy committed:
Changed observation space channel dimension to be first

More regular for PyTorch and therefore in general

fernandotorch commented on algorithms/a3c/test.py:
```diff
@@ -0,0 +1,78 @@
+
+# from envs import create_atari_env
+from gym_ai2thor.envs.ai2thor_env import AI2ThorEnv
+from model import ActorCritic
```
Change the imports relative to root path. 
```suggestion
from algorithms.a3c.model import ActorCritic
```
This applies also in other scripts

beduffy replied:
is 
`from test import test`
wrong too?
Where do you draw the line? Not everything is the full path? relative paths are ok when?

beduffy replied:
`from algorithms.a3c.test import test`
I did this but pycharm gives warning:
`package containing module 'algorithms' is not listed in project requirements`

fernandotorch commented on algorithms/a3c/envs.py:
```diff
@@ -0,0 +1,59 @@
+    frame = frame[34:34 + 160, :160]
+    # Resize by half, then down to 42x42 (essentially mipmapping). If
+    # we resize directly we lose pixels that, when mapped to 42x42,
+    # aren't close enough to the pixel boundary.
```
This 3 commented lines should probably be in the docstring of _process_frame42. Otherwise it breaks the flow of the code

beduffy replied:
meh, it's not my code, might keep it that way

beduffy committed:
Added atari arg params so we can swap between any atari and AI2ThorEnv

Only render atari in test.

beduffy committed:
Added function to calculate input size to lstm and general reshape
operation needed after 4 conv layers for square image. Renamed frame_width to frame_dim

beduffy committed:
Fixed 2 massive PutObject bugs! Closed AI2ThorEnv properly. More.

Added interaction_open_close config key to remove open/close actions if needed

beduffy committed:
Added to README e.g. ikostrikov, paths, better example, typos.

Extra todo, atari render default.

fernandotorch commented on README.md:
```diff
@@ -19,8 +19,10 @@ More detailed information on ai2thor environment can be found on their
 - [Gated-Attention Architectures for Task-Oriented Language Grounding](https://arxiv.org/abs/1706.07230) 
--- *Original code available on [DeepRL-Grounding](https://github.com/devendrachaplot/DeepRL-Grounding)*
+-- *Original code available on [DeepRL-Grounding](https://github.com/devendrachaplot/DeepRL-Grounding)* 
+and happens to use Ikostrikov's A3C as well
```
```suggestion
also based on Ikostrikov's A3C
```

fernandotorch commented on README.md:
```diff
@@ -60,13 +62,14 @@ for episode in range(N_EPISODES):
-# envs/configs/myconfig.json
+# gym_ai2thor/config_files/myconfig.json
 {'env': {'interaction': True,
+         'interaction_open_close': true,
```
What we discussed about interaction hierarchy

beduffy replied:
`interaction_pickup_put`
to replace interaction?
We didn't finalize the names, `interaction` alone as so clear

fernandotorch replied:
Disagree. If you separate open/close and are explicit, then you need to be explicit for the other as well

fernandotorch commented on README.md:
```diff
@@ -103,19 +106,19 @@ class MoveAheadTask(BaseTask):
-        reward = -1 if state.metadata['lastAction'] == 'MoveAhead' else 1 
-        done = reward > 100 or self.max_episode_length
+        reward = 1 if state.metadata['lastAction'] == 'MoveAhead' else -1 
+        done = reward > 100 or self.step_num > self.max_episode_length
```
We should add self.reward to keep track of rewards and use that, otherwise it is impossible to reach or surpass even a value of reward 2

beduffy replied:
`self.rewards.append(reward)`
`done = sum(rewards) > 100 or self.step_num > self.max_episode_length`
`if done:`
 `           self.rewards = []`

fernandotorch commented on algorithms/a3c/main.py:
```diff
@@ -0,0 +1,116 @@
+    # TODO resolve if env wrapper/factory is needed instead of env if statements
+    # TODO gym ai2thor changes (channel first (done) and remove env in config?)
+    # TODO README.md
+    # TODO docstrings at the top of the file and import problem?
```
Please remove todos before PR

fernandotorch commented on algorithms/a3c/main.py:
```diff
@@ -0,0 +1,116 @@
+"""
+Adapted from: https://github.com/ikostrikov/pytorch-a3c/blob/master/main.py
+"""
```
I need more information about what this code does in the docstring here

fernandotorch commented on algorithms/a3c/envs.py:
```diff
@@ -0,0 +1,59 @@
+"""
+Adapted from: https://github.com/ikostrikov/pytorch-a3c/blob/master/train.py
+"""
```
More information on the docstring, something like "auxiliary wrappers from openAI gym environment"

fernandotorch commented on algorithms/a3c/model.py:
```diff
@@ -0,0 +1,94 @@
+"""
+Adapted from: https://github.com/ikostrikov/pytorch-a3c/blob/master/model.py
+"""
```
Again, please describe what this script is for

fernandotorch commented on algorithms/a3c/model.py:
```diff
@@ -0,0 +1,94 @@
+    42x42 -> (42 − 3 + 2)÷ 2 + 1 = 21 after 1 layer
+    11 after 2 -> 6 -> and finally width 3
+    Therefore lstm input size would be (3 * 3 * num_filters)
+    """
```
Love it

fernandotorch commented on algorithms/a3c/model.py:
```diff
@@ -0,0 +1,94 @@
+    return width * width * num_filters
+
+def normalized_columns_initializer(weights, std=1.0):
+    out = torch.randn(weights.size())
```
Why or when do I need this function? It would be good to have a one line docstring

fernandotorch commented on algorithms/a3c/model.py:
```diff
@@ -0,0 +1,94 @@
+
+
+class ActorCritic(torch.nn.Module):
+    def __init__(self, num_inputs, action_space, frame_dim):
```
Please describe in docstring this class. You have lstms and shit, so it is not a typical actor critic that doesn't need explanation. Also reference to paper

fernandotorch commented on algorithms/a3c/my_optim.py:
```diff
@@ -0,0 +1,76 @@
+"""
+Adapted from: https://github.com/ikostrikov/pytorch-a3c/blob/master/my_optim.py
+"""
```
Describe what the script is for in the docstring

fernandotorch commented on algorithms/a3c/test.py:
```diff
@@ -0,0 +1,83 @@
+"""
+Adapted from https://github.com/ikostrikov/pytorch-a3c
+"""
```
More information in the docstring

fernandotorch commented on algorithms/a3c/test.py:
```diff
@@ -0,0 +1,83 @@
+        # Sync with the shared model
+        if done:
+            model.load_state_dict(shared_model.state_dict())
+            cx = Variable(torch.zeros(1, 256), volatile=True)
```
Variable is deprecated. I suggest we change it

fernandotorch commented on algorithms/a3c/test.py:
```diff
@@ -0,0 +1,83 @@
+
+        value, logit, (hx, cx) = model((Variable(state.unsqueeze(0).float(), volatile=True),
+                                        (hx, cx)))
+        prob = F.softmax(logit)
```
This use of softmax is deprecated, give the dim=1

beduffy replied:
`dim=-1`

fernandotorch commented on algorithms/a3c/train.py:
```diff
@@ -0,0 +1,145 @@
+"""
+Adapted from https://github.com/ikostrikov/pytorch-a3c
+"""
```
Docstring not explaining script

fernandotorch commented on gym_ai2thor/config_files/config_example.json:
```diff
@@ -1,10 +1,9 @@
 {
     "env": {
         "interaction": true,
+        "interaction_open_close": true,
```
As discussed

beduffy committed:
Added detailed docstrings to 6 of ikostrikov's files and ActorCritic.

Added example of how to run a3c in main README and better task example.
Added full root path to algorithms and set sync default right.

beduffy committed:
Improved, renamed cup picking test case to try other actions and interactions

Couldn't put cup back to where it was so decided to put it in the Microwave

fernandotorch commented on algorithms/a3c/envs.py:
```diff
@@ -0,0 +1,62 @@
+"""
+Adapted from: https://github.com/ikostrikov/pytorch-a3c/blob/master/envs.py
```
I would put adapted below the paragraph that starts with "This contains". First explain what it is, then go through the details, i.e. giving credit (of course within the same docstring as you did)

fernandotorch commented on algorithms/a3c/envs.py:
```diff
@@ -0,0 +1,62 @@
+from gym.spaces.box import Box
+
+
+# Taken from https://github.com/openai/universe-starter-agent
```
```suggestion
# Original code from https://github.com/openai/universe-starter-agent
```

beduffy replied:
Again, not my code so not changing this part

fernandotorch commented on algorithms/a3c/main.py:
```diff
@@ -0,0 +1,124 @@
+"""
+Adapted from: https://github.com/ikostrikov/pytorch-a3c/blob/master/main.py
```
Add space and adapted below the explanation

fernandotorch commented on algorithms/a3c/main.py:
```diff
@@ -0,0 +1,124 @@
+"""
+Adapted from: https://github.com/ikostrikov/pytorch-a3c/blob/master/main.py
+This is the main file needed within a3c and runs of the train and test functions from
```
```suggestion
Main file needed within a3c. Runs the train and test functions from
```

fernandotorch commented on algorithms/a3c/main.py:
```diff
@@ -0,0 +1,124 @@
+Adapted from: https://github.com/ikostrikov/pytorch-a3c/blob/master/main.py
+This is the main file needed within a3c and runs of the train and test functions from
+their respective files.
+Example use case:
```
```suggestion
Example of use:
```

fernandotorch commented on algorithms/a3c/main.py:
```diff
@@ -0,0 +1,124 @@
+`cd algorithms/a3c`
+`python main.py`
+
+This runs A3C on our AI2ThorEnv wrapper with default params (4 processes). Optionally it can be
```
```suggestion
Runs A3C on our AI2ThorEnv wrapper with default params (4 processes). Optionally it can be
```

fernandotorch commented on algorithms/a3c/main.py:
```diff
@@ -0,0 +1,124 @@
+`python main.py`
+
+This runs A3C on our AI2ThorEnv wrapper with default params (4 processes). Optionally it can be
+run on any atari environment as well with the --atari and --atari-env-name params.
```
```suggestion
run on any atari environment as well using the --atari and --atari-env-name params.
```

fernandotorch commented on algorithms/a3c/main.py:
```diff
@@ -0,0 +1,124 @@
+
+    # TODO resolve if env wrapper/factory is needed instead of env if statements
+    # TODO gym ai2thor changes (channel first (done) and remove env in config?)
+    # TODO import problem?
```
TODOS out

beduffy replied:
waiting for remove env reply

fernandotorch commented on algorithms/a3c/model.py:
```diff
@@ -0,0 +1,108 @@
+"""
+Adapted from: https://github.com/ikostrikov/pytorch-a3c/blob/master/model.py
```
Adapted below

fernandotorch commented on algorithms/a3c/model.py:
```diff
@@ -0,0 +1,108 @@
+"""
+Adapted from: https://github.com/ikostrikov/pytorch-a3c/blob/master/model.py
+
+This code contains the main A3C model which outputs predicted value, action logits and hidden state.
```
```suggestion
Main A3C model which outputs predicted value, action logits and hidden state.
```

fernandotorch commented on algorithms/a3c/model.py:
```diff
@@ -0,0 +1,108 @@
+Adapted from: https://github.com/ikostrikov/pytorch-a3c/blob/master/model.py
+
+This code contains the main A3C model which outputs predicted value, action logits and hidden state.
+Some helper functions too for weight initialisation and dynamically computing LSTM/flatten input
```
```suggestion
Includes helper functions for weight initialisation and dynamically computing LSTM/flatten input
```

fernandotorch commented on algorithms/a3c/model.py:
```diff
@@ -0,0 +1,108 @@
+
+class ActorCritic(torch.nn.Module):
+    """
+    Mainly Ikostrikov's implementation of A3C (https://arxiv.org/abs/1602.01783).
```
One blank space after this line

fernandotorch commented on algorithms/a3c/model.py:
```diff
@@ -0,0 +1,108 @@
+class ActorCritic(torch.nn.Module):
+    """
+    Mainly Ikostrikov's implementation of A3C (https://arxiv.org/abs/1602.01783).
+    The algorithm processes an input image (with num_input_channels) with 4 conv layers,
```
```suggestion
    Processes an input image (with num_input_channels) with 4 conv layers,
```

beduffy committed:
Removed all deprecation warnings and changed to torch 0.4+

softmax -1 dim spec, no variable, more detach, clip_grad_norm_
skimage mode='reflect' will be default in future
Also changed all imports to be from root path

fernandotorch commented on algorithms/a3c/my_optim.py:
```diff
@@ -0,0 +1,82 @@
+"""
+Adapted from: https://github.com/ikostrikov/pytorch-a3c/blob/master/my_optim.py
```
Adapted below

fernandotorch commented on algorithms/a3c/my_optim.py:
```diff
@@ -0,0 +1,82 @@
+"""
+Adapted from: https://github.com/ikostrikov/pytorch-a3c/blob/master/my_optim.py
+
+In the original A3C paper (https://arxiv.org/abs/1602.01783) the authors compared 3 different
```
```suggestion
In the original A3C paper (https://arxiv.org/abs/1602.01783), the authors compared 3 different
```

fernandotorch commented on algorithms/a3c/my_optim.py:
```diff
@@ -0,0 +1,82 @@
+
+In the original A3C paper (https://arxiv.org/abs/1602.01783) the authors compared 3 different
+optimizers i.e. Momentum SGD, RMSProp and Shared RMSProp (check final part of section 4). The
+difference between the 3rd compared to the 2nd was whether to compute shared statistics across all
```
```suggestion
difference between the 3rd compared to the 2nd is whether to compute shared statistics across all
```

fernandotorch commented on algorithms/a3c/my_optim.py:
```diff
@@ -0,0 +1,82 @@
+In the original A3C paper (https://arxiv.org/abs/1602.01783) the authors compared 3 different
+optimizers i.e. Momentum SGD, RMSProp and Shared RMSProp (check final part of section 4). The
+difference between the 3rd compared to the 2nd was whether to compute shared statistics across all
+threads and this was found to be more robust. It seems the equivalent was implemented for Adam
```
```suggestion
threads, which was found to be more robust. It seems the equivalent was implemented for Adam
```

fernandotorch commented on algorithms/a3c/test.py:
```diff
@@ -0,0 +1,87 @@
+"""
+Adapted from https://github.com/ikostrikov/pytorch-a3c/blob/master/test.py
```
Adapted below

fernandotorch commented on algorithms/a3c/test.py:
```diff
@@ -0,0 +1,87 @@
+"""
+Adapted from https://github.com/ikostrikov/pytorch-a3c/blob/master/test.py
+
+This code contains the testing loop of the shared model within A3C (no optimisation/backprop needed)
```
```suggestion
Contains the testing loop of the shared model within A3C (no optimisation/backprop needed)
```

fernandotorch commented on algorithms/a3c/train.py:
```diff
@@ -0,0 +1,156 @@
+"""
+Adapted from https://github.com/ikostrikov/pytorch-a3c/blob/master/train.py
```
Below

fernandotorch commented on algorithms/a3c/train.py:
```diff
@@ -0,0 +1,156 @@
+"""
+Adapted from https://github.com/ikostrikov/pytorch-a3c/blob/master/train.py
+
+This contains the train code run by each A3C process on either Atari or AI2ThorEnv.
```
```suggestion
Contains the train code run by each A3C process on either Atari or AI2ThorEnv.
```

beduffy committed:
Added Fernando's grammar corrections to docstrings and renamed action_space

beduffy committed:
Removed env key in config for UX, fixed bug of task not knowing objects

renamed interaction keys, inventory changed str only when changed.
Complex config reading but worth it.
Showed two ways in task_variation_random_walk.py of disabling opening.
Docstring to normalized_columns_init

beduffy committed:
Removed final TODO and fixed up normalized_init docstring
